### PR TITLE
Denormalization should consider the LHS of MOVEs

### PR DESCRIPTION
**Pull request by benharsh**
While working on another branch I observed that denormalization could
transform the following AST:

(move T (+ A B))
(move A B)
(move B T)

into this incorrect pattern:

(move A B)
(move B (+ A B))

It turned out that denormalization wasn't recognizing the 'move A B'
call inbetween the first and last expressions.

This does not appear to have a significant impact on denormalization's
effectiveness at this time.

Based on patch received from @e-kayrakli 

Successfully passed the following testing configs:
- [x] --local
- [x] --no-local
- [x] gasnet

### CODE REVIEW

File: compiler/util/exprAnalysis.cpp [hunk 1]
```diff
@@ -33,7 +33,7 @@
  * could distinguish between reads and writes to memory and to take into
  * account alias analysis.
  */
-bool SafeExprAnalysis::exprHasNoSideEffects(Expr* e) {
+bool SafeExprAnalysis::exprHasNoSideEffects(Expr* e, Expr* exprToMove) {
```
Comment (mppf) [inline compiler/util/exprAnalysis.cpp]: You need to update the comment above about what this function checks

File: compiler/util/exprAnalysis.cpp [hunk 2]
```diff
@@ -54,6 +54,33 @@ bool SafeExprAnalysis::exprHasNoSideEffects(Expr* e) {
...
+          INT_ASSERT(isSymExpr(ce->get(1)));
+          std::vector<SymExpr*> syms;
+          collectSymExprs(exprToMove, syms);
+          for_vector(SymExpr, s, syms) {
+            if (s->var == toSymExpr(ce->get(1))->var) {
+              safeExprCache[e] = false;
+              return false;
+            }
+          }
+        }
```
Comment (mppf) [inline compiler/util/exprAnalysis.cpp]: couldn't this also happen through a reference? Or have the variable set with PRIM_ASSIGN?
I generally think this is the right place for the check, I'm just not sure it covers all the possible cases.

**Comment by bradcray** [comment]
Did catching this so late in the release cycle give anyone the desire to turn of denormalization by default for this release and to turn it on shortly after it goes out to live with it longer?  Or was the issue only triggered by changes in the compiler made since then?

**Comment by mppf** [comment]
There are potentially other related issues (and I think Ben was planning to make a PR addressing them) but we've never seen this issue on master. Ben noticed the problem on the branch adjusting blank intent (which makes a bunch of reference-related changes and uses QualifiedType).

**Comment by benharsh** [comment]
Michael is correct. I believe the problem on my dev branch arose from new/different uses of PRIM_MOVE and PRIM_ASSIGN. The fix seemed simple enough that it could go into master, and could only solidify denormalization's correctness.

The additional case I would like to address is the pattern `(move A B)` where `A` is a reference and `B` is not. This *should* use PRIM_ASSIGN, but I'm not convinced we're using the two correctly on master.

There are possibly other issues, but there may not be time to address them and I'm hoping the upcoming ref changes will make it easier to identify problems.

**Pull request merged**